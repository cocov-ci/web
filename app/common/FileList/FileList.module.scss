@use "sass:map";
@import "styles/variables.module";

.base {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;

  .header {
    display: flex;
    flex-direction: row;
    border-bottom: 1px solid $colorGray;

    > div {
      padding: 9px 15px;
    }

    .coverage {
      width: 265px;
    }

    .file {
      flex: 1;
    }
  }

  .fileList {
    overflow-y: scroll;
    flex: 1;
    display: flex;
    flex-direction: column;

    .fileItem {
      display: flex;
      border: 1px solid transparent;

      &:nth-child(odd) {
        background-color: map.get($colorFileList, "oddItemBackground");
      }

      &:hover {
        background: map.get($colorFileList, "hoverItemBackground");
        border: 1px solid map.get($colorFileList, "hoverItemBorder");
        border-radius: 8px;
      }

      > .file,
      .percentage {
        padding: 6px 15px;
      }

      .file {
        flex: 1;
        display: flex;
        align-items: center;
      }

      .percentage {
        display: flex;
        align-items: center;
        width: 265px;
        border-left: 1px solid $colorGray;

        .percentageValue {
          text-align: right;
          width: 55px;
          margin-right: 6px;
        }

        .percentageProgressBar {
          width: 200px;
        }
      }
    }
  }
}
