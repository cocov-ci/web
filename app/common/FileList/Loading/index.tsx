'use client'

import classNames from 'classnames'
import React from 'react'

import Loading from 'app/common/Loading'

import styles from '../FileList.module.scss'

export type CommitHeaderProps = {
  className?: string
}

const LoadingComponent = ({ className }: CommitHeaderProps) => {
  return (
    <div className={classNames(styles.base, className)}>
      <div className={styles.header}>
        <div className={styles.file}>
          <Loading
            alignment="left"
            height="24px"
            type="skeleton"
            width="100px"
          />
        </div>
        <div className={styles.coverage}>
          <Loading
            alignment="left"
            height="24px"
            type="skeleton"
            width="100px"
          />
        </div>
      </div>

      <div className={styles.fileList}>
        {Array.from(Array(11).keys()).map(item => (
          <div key={item}>
            <div className={styles.fileItem}>
              <Loading height="28px" type="skeleton" width="100%" />
            </div>
            <div className={styles.fileItem} />
          </div>
        ))}
      </div>
    </div>
  )
}

export default LoadingComponent
