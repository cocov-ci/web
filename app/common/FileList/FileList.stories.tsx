import type { Meta, StoryObj } from '@storybook/react'

import FileList from './index'

const meta: Meta<typeof FileList> = {
  title: 'Common/FileList',
  component: FileList,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof FileList>

export const Default: Story = {
  args: {
    files: [
      {
        id: 1,
        file: 'app/models/application_record.rb',
        percentCovered: 57.14,
      },
      { id: 2, file: 'app/lib/cocov/git_hub.rb', percentCovered: 90.48 },
      { id: 3, file: 'app/models/private_key.rb', percentCovered: 91.43 },
      { id: 4, file: 'app/models/user.rb', percentCovered: 92.86 },
      {
        id: 5,
        file: 'app/services/git_service/base_storage.rb',
        percentCovered: 95.24,
      },
      { id: 6, file: 'app/models/user_token.rb', percentCovered: 95.83 },
      { id: 7, file: 'app/lib/cocov/crypto.rb', percentCovered: 96.97 },
      {
        id: 8,
        file: 'app/lib/cocov/schema_validator.rb',
        percentCovered: 98.25,
      },
      {
        id: 9,
        file: 'app/controller/application_controller.rb',
        percentCovered: 100,
      },
    ],
  },
}
