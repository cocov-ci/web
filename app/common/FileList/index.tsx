'use client'

import classNames from 'classnames'
import Link from 'next/link'
import { useRouter } from 'next/navigation'
import React, { useEffect, useRef } from 'react'

import { CoverageFileProps } from 'types/Coverage'

import ProgressBar from '../ProgressBar'

import styles from './FileList.module.scss'
import Loading from './Loading'

type FileListProps = {
  className?: string
  loading?: boolean
  resetScrollAfterPageLoad?: boolean
  commitSha: string
  repositoryName: string
  files: CoverageFileProps[]
}

const FileList = ({
  className,
  files,
  commitSha,
  loading,
  repositoryName,
  resetScrollAfterPageLoad,
}: FileListProps) => {
  const listRef = useRef<HTMLDivElement>(null)

  const onClick = () => {
    const list = listRef.current

    if (list) {
      localStorage.setItem(
        `coverage-list-${commitSha}`,
        list.scrollTop?.toString(),
      )
    }
  }

  useEffect(() => {
    const top = localStorage.getItem(`coverage-list-${commitSha}`)
    const list = listRef.current

    if (list && top !== null) {
      list.scrollTop = parseInt(top, 10)
    }

    if (resetScrollAfterPageLoad) {
      localStorage.removeItem(`coverage-list-${commitSha}`)
    }
  }, [])

  if (loading) return <Loading className={className} />

  return (
    <div className={classNames(styles.base, className)}>
      <div className={styles.header}>
        <div className={styles.file}>File</div>
        <div className={styles.coverage}>Coverage</div>
      </div>
      <div className={styles.fileList} ref={listRef}>
        {files?.map(f => (
          <Link
            className={styles.fileItem}
            href={`/repos/${repositoryName}/commits/${commitSha}/coverage/${f.id}`}
            key={f.id}
            onClick={() => onClick()}
          >
            <div className={styles.file}>{f.file}</div>
            <div className={styles.percentage}>
              <div className={styles.percentageValue}>{f.percent_covered}%</div>
              <div className={styles.percentageProgressBar}>
                <ProgressBar value={f.percent_covered} width="100%" />
              </div>
            </div>
          </Link>
        ))}
      </div>
    </div>
  )
}

export default FileList
