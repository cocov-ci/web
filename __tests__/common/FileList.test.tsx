import { fireEvent, render, screen } from '@testing-library/react'

import FileList from 'app/common/FileList'

import '@testing-library/jest-dom'

const args = {
  commitSha: '1234356',
  repositoryName: 'api',
  files: [
    {
      id: 1,
      file: 'app/models/application_record.rb',
      percent_covered: 57.14,
    },
    { id: 2, file: 'app/lib/cocov/git_hub.rb', percent_covered: 90.48 },
  ],
}

describe('common/FileList', () => {
  it('renders FileList', () => {
    const { container } = render(<FileList {...args} />)

    expect(container.querySelectorAll('.fileList a').length).toEqual(2)
    expect(screen.getByText('57.14%')).toBeVisible()
    expect(screen.getByText('app/models/application_record.rb')).toBeVisible()
  })

  it('triggers onClick', () => {
    render(<FileList {...args} resetScrollAfterPageLoad />)

    fireEvent.click(screen.getByText('app/models/application_record.rb'))
  })

  it('renders FileList with empty list', () => {
    const { container } = render(
      <FileList commitSha="1234356" files={[]} repositoryName="api" />,
    )

    expect(container.querySelectorAll('.base').length).toEqual(0)
  })

  it('renders FileList with "loading" param', () => {
    const { container } = render(<FileList {...args} loading={true} />)

    expect(
      container.getElementsByClassName('react-loading-skeleton').length,
    ).toEqual(13)
  })

  it('renders FileList snapshots', () => {
    const { container } = render(<FileList {...args} />)

    expect(container).toMatchSnapshot()
  })
})
