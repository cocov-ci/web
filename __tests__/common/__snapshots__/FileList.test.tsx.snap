// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/FileList renders FileList snapshots 1`] = `
<div>
  <div
    class="base"
  >
    <div
      class="header"
    >
      <div
        class="file"
      >
        File
      </div>
      <div
        class="coverage"
      >
        Coverage
      </div>
    </div>
    <div
      class="fileList"
    >
      <a
        class="fileItem"
        href="/repos/api/commits/1234356/coverage/1"
      >
        <div
          class="file"
        >
          app/models/application_record.rb
        </div>
        <div
          class="percentage"
        >
          <div
            class="percentageValue"
          >
            57.14
            %
          </div>
          <div
            class="percentageProgressBar"
          >
            <div
              class="base"
              style="width: 100%; height: 7px;"
            >
              <div
                class="progressFill"
                style="width: 57.14%;"
              />
            </div>
          </div>
        </div>
      </a>
      <a
        class="fileItem"
        href="/repos/api/commits/1234356/coverage/2"
      >
        <div
          class="file"
        >
          app/lib/cocov/git_hub.rb
        </div>
        <div
          class="percentage"
        >
          <div
            class="percentageValue"
          >
            90.48
            %
          </div>
          <div
            class="percentageProgressBar"
          >
            <div
              class="base"
              style="width: 100%; height: 7px;"
            >
              <div
                class="progressFill"
                style="width: 90.48%;"
              />
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
